Changes for preventing exceptions on failovers

charkost opened the pull request:
### Restart monitoring threads on child processes if needed

After mongodb server failovers, child processes are unable to detect
the new primary server due to the monitoring threads' absence:

```
No primary server is available in cluster: #<Cluster topology=ReplicaSetNoPrimary[192.168.22.162:27017,192.168.22.254:27017,192.168.22.161:27017,name=rs0,v=49,e=7fffffff0000000000000056] servers=[#<Server address=192.168.22.162:27017 UNKNOWN pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>,#<Server address=192.168.22.254:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>,#<Server address=192.168.22.161:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=1 (0-5) used=0 avail=1 pending=0>>]> with timeout=30, LT=0.015. The following servers have dead monitor threads: #<Server address=192.168.22.162:27017 UNKNOWN pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>, #<Server address=192.168.22.254:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>, #<Server address=192.168.22.161:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=1 (0-5) used=0 avail=1 pending=0>> (Mongo::Error::NoServerAvailable)
```

This error can be prevented using `cluster.reconnect!` when server selection fails
due to dead monitoring threads on child processes. Monitoring threads
start & perform the necessary topology update resulting in server selection
success.

NOTE: This is a temporary workaround till our resque forking model is
changed to avoid forking for each job execution. Using the driver from
child processes with absent monitoring threads is officially discouraged.

### Forwardport ping check before every operation
The latest driver version is unable to handle some writes in the legacy codepath resulting in `Errno::EPIPE: Broken` pipe exceptions.

These exceptions arise due to outdated topology state in the short time window between the monitor thread scans & under the following conditions:
    


 - Failover is performed, the previous primary server becomes secondary & closes the connection with the driver.
 - The connection from the driver's side changes to CLOSE_WAIT state but the driver is clueless about that until the next monitor thread scan.
 - At least 2 writes are performed on the outdated connection. On the second write Errno::EPIPE: Broken pipe is raised.
    
While the same error is possible for reads as well, the driver retries only the read calls propably due to the inability for bulk-write retries.
    
In contrary to the latest driver, the previous driver versions are able to handle those exceptions by sending a ping before each operation. Pings rescue all exceptions & topology update is performed after a failed ping.

As a workaround for this problem, this ping is re-introduced in the latest driver version.

NOTE: The modern codepath which is enabled on server versions >=3.6 is able to handle those exceptions without pinging.

charkost committed:
Restart monitor threads upon server selection failure

After mongodb server failovers, child processes are unable to detect
the new primary server due to the monitoring threads' absence:

```
No primary server is available in cluster: #<Cluster topology=ReplicaSetNoPrimary[192.168.22.162:27017,192.168.22.254:27017,192.168.22.161:27017,name=rs0,v=49,e=7fffffff0000000000000056] servers=[#<Server address=192.168.22.162:27017 UNKNOWN pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>,#<Server address=192.168.22.254:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>,#<Server address=192.168.22.161:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=1 (0-5) used=0 avail=1 pending=0>>]> with timeout=30, LT=0.015. The following servers have dead monitor threads: #<Server address=192.168.22.162:27017 UNKNOWN pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>, #<Server address=192.168.22.254:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=0 (0-5) used=0 avail=0 pending=0>>, #<Server address=192.168.22.161:27017 SECONDARY replica_set=rs0 pool=#<ConnectionPool size=1 (0-5) used=0 avail=1 pending=0>> (Mongo::Error::NoServerAvailable)
```

This error can be prevented using `cluster.reconnect!` when server selection fails
due to dead monitoring threads on child processes. Monitoring threads
start & perform the necessary topology update resulting in server selection
success.

NOTE: This is a temporary workaround till our resque forking model is
changed to avoid forking for each job execution. Using the driver from
child processes with absent monitoring threads is officially discouraged.

charkost committed:
Temporarily allow parent-initiated connections

This is a temporary workaround till our resque forking model is
changed to avoid forking for each job execution. Using the driver from
child processes with parent-initiated connections is officially discouraged.

agis commented:
Is there any chance this could be approved upstream? Do you think it's worth trying? (Maybe we can get some good feedback this way.)

charkost committed:
Forwardport ping check before every operation

The latest driver version is unable to handle some writes in the legacy
codepath resulting in Errno::EPIPE: Broken pipe exceptions.

These exceptions arise due to outdated topology state in the short
time window between the monitor thread scans & under the following
conditions:

 - Failover is performed, the previous primary server becomes secondary
   & closes the connection with the driver.
 - The connection from the driver's side changes to CLOSE_WAIT state but
   the driver is clueless about that until the next monitor thread scan.
 - At least 2 writes are performed on the outdated connection.
   On the second write Errno::EPIPE: Broken pipe is raised.

While the same error is possible for reads as well, the driver retries
only the read calls propably due to the inability for bulk-write retries.

In contrary to the latest driver, the previous driver versions are able
to handle those exceptions by sending a ping before each operation.
Pings rescue all exceptions & topology update is performed after a failed ping.

As a workaround for this problem, this ping is re-introduced in the
latest driver version.

NOTE: The modern codepath which is enabled on server versions >=3.6 is
able to handle those exceptions without pinging.

charkost committed:
Bump version to 2.13.0.skroutz.1
